core/src/main/java/hudson/model/AbstractBuild.java

```diff
@@ -459,49 +460,64 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
             hudsonVersion = Jenkins.VERSION;
             this.listener = listener;
 
-            launcher = createLauncher(listener);
-            if (!Jenkins.get().getNodes().isEmpty()) {
-                if (node instanceof Jenkins) {
-                    listener.getLogger().print(Messages.AbstractBuild_BuildingOnMaster());
-                } else {
-                    listener.getLogger().print(Messages.AbstractBuild_BuildingRemotely(ModelHyperlinkNote.encodeTo("/computer/" + builtOn, node.getDisplayName())));
-                    Set<LabelAtom> assignedLabels = new HashSet<>(node.getAssignedLabels());
-                    assignedLabels.remove(node.getSelfLabel());
-                    if (!assignedLabels.isEmpty()) {
-                        boolean first = true;
-                        for (LabelAtom label : assignedLabels) {
-                            if (first) {
-                                listener.getLogger().print(" (");
-                                first = false;
-                            } else {
-                                listener.getLogger().print(' ');
+            Result result = null;
+            buildEnvironments = new ArrayList<>();
```

Comment:
This historical implementation just asks for concurrency issues at some point. Not blocking this change, because it does not make it any worse

Reply:
> I do not see any implementation issues except the loss of information when `InterruptedException `is thrown by the teardown logic. It looks like the results should be combined taking handling of other negative scenarios. The rest of the code looks good to me.

@oleg-nenashev: Are you ok with https://github.com/jenkinsci/jenkins/pull/4517#discussion_r389186582 or are you still requesting changes in the handling of `InterruptedException` from teardown?

One minor, but simple, improvement I can think of would be https://github.com/thomasgl-orange/jenkins/commit/2e9b34260a4000d12f0b769324c2d6499d457aec to log the `Result` of the `doRun` phase, which is the information we're about to lose when re-throwing the `InterruptedException`. Would that help, or is there something else you had in mind?